src/io/data_file_writer.rs

```diff
@@ -0,0 +1,275 @@
+//! data_file is used create a data file writer to write data into files.
+
+use std::collections::HashMap;
+
+use crate::{types::DataFile, Result};
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use opendal::Operator;
+use parquet::format::FileMetaData;
+
+use super::{
+    location_generator::DataFileLocationGenerator,
+    parquet::{ParquetWriter, ParquetWriterBuilder},
+};
+
+/// A writer capable of splitting incoming data into multiple files within one spec/partition based on the target file size.
+/// When complete, it will return a list of `DataFile`.
+pub struct DataFileWriter {
+    operator: Operator,
+    location_generator: DataFileLocationGenerator,
+    arrow_schema: SchemaRef,
+
+    rows_divisor: usize,
+    /// # TODO
+    ///
+    /// Using `target_file_size_in_bytes`. The `ParquetWriter` can't support to track file size in bytes now.
+    target_file_row_num: usize,
+
+    /// # TODO
+    ///
+    /// support to config `ParquetWriter` using `buffer_size` and writer properties.
+    current_writer: Option<ParquetWriter>,
+    current_row_num: usize,
+    /// `current_location` used to clean up the file when no row is written to it.
+    current_location: String,
+
+    result: Vec<DataFile>,
+}
+
+impl DataFileWriter {
+    /// Create a new `DataFileWriter`.
+    pub async fn try_new(
+        operator: Operator,
+        location_generator: DataFileLocationGenerator,
+        arrow_schema: SchemaRef,
+        rows_divisor: usize,
+        target_file_row_num: usize,
+    ) -> Result<Self> {
+        let mut writer = Self {
+            operator,
+            location_generator,
+            arrow_schema,
+            rows_divisor,
+            target_file_row_num,
+            current_writer: None,
+            current_row_num: 0,
+            current_location: String::new(),
+            result: vec![],
+        };
+        writer.open_new_writer().await?;
+        Ok(writer)
+    }
+
+    /// Write a record batch. The `DataFileWriter` will create a new file when the current row num is greater than `target_file_row_num`.
+    pub async fn write(&mut self, batch: RecordBatch) -> Result<()> {
```

Comment:
I don't like the API to do `proxy`. How about implementing API like the  following:

For ParquetWriter:

- `ParquetWriter::write()` should return `num_rows`.

For DataFileWriter:

- `DataFileWriter::next(&mut self) -> (ParquetWriterContext, ParquetWriterBuilder)`: start a new file write, users can configure the wirter by builder. `ParquetWriterContext` carries context like `rows_divisor` and `target_file_row_num`
- `DataFileWriter::commit(&self, ParquetWriterContext, FileMeta) -> DataFile`:  Commit written data into DataFile with all info.

---

In this way:

- We don't need to track and hold the state in `DataFileWriter`.
- We don't need to inject all `ParquetWriter` related options in `DataFileWriter`.
- We can write data file in concurrent way.
- Callers can decide how to use or track `DataFile`.
- Easier to support more formats (in the future).

Reply:
Sounds a good design for me. Does this means that the user should check using context when it writes like following:
```
let (context,writer_builder) = DataFileWriter::next(&mut self);
let writer = writer_builder.build();
let mut cur_row_num = 0;
while let cur_row_num < context.target_file_row_num { <---- // User should check by self 
    let row_num = writer.write();
    cur_row_num = cur_row_num + row_num;
}
```

Reply:
Let's polish the API in following PRs.